### DISCUSSION TITLE
wz_mini.conf erased when using web cgi script to 'revert' a V3 config change

### DISCUSSION THREAD

**Discussion by Bob-AOL** [gtxaspec/wz_mini_hacks#249] [General]
Title says it all. I simply made a frame rate adjustment via the wz_mini web cgi script that didn't play nice with Blue Iris. Went back and used 'revert' to the original (and only other) config file instead of actually editing and 'refreshing the script. Tried reboot from the script which rarely works and didn't this time either. Rebooted the camera from the app instead. When it came back online it was no longer accessible from port 80 to review the changes. A scan revealed it was no longer connected via the USB interface via ethernet (which it was previously) and had reverted to wifi with a new ip address and only port 22 and port 554 open -- the base 4.61.0.3 RTSP firmware.
Climbed up and retrieved the memory card to find both conf files blank and an additional phantom wz_mini.conf file with no designation and also no size -- presumably also with no contents.

Obviously a bug with significance consequences. 

The good news is, a hardline ethernet connection dramatically increases bitrate well beyond the typical 100~130 kB/s to upwards of 250kB/s -- which is 25fps territory and why I tried changing the frame rate. Blue Iris was already reporting 22fps for this particular camera which was set for 20. I figured what the hell, let's see if it'll lock in at 25 even though it's supposed to be software limited. Nope Made the change in both BI and the wz_mini.conf before rebooting, btw. All it did was break things.

Forewarned is forearmed.

**Comment by virmaior** [comment]
I just saw this now.
 
What version of the wz_mini hack were/are you running? What type of camera is this?
Also isn't it using the stock RTSP firmware advised against for this?

1.  reboot from the script wasn't something that "rarely works". It either works or doesn't (it's not a buggy but functional item). and was broken due to a coding error. the reboot issue is fixed (or should be?) in the latest version as noted in the other discussion you posted from.
2.  I just tried update/revert with the latest version from the web interface and was unable to duplicate the issue you described.

Here's what I just tested as functioning:
```
1. modified a field
2. clicked update
3. made a backup in case it does explode as you described
4. clicked on 1 down at the bottom
5. clicked on revert at the top while the url was /cgi-bin/config.cgi?action=show_revert&version=1
6. checked the directory and the file was correctly there ...
```
 Can you describe in more detail what you did?

I could imagine it being buggy if you double clicked at either the update or revert phase or something of that sort or if the SD card is dying. Could also be cases that are not being covered.

**Comment by Bob-AOL** [comment]
Cam was a V3 running the latest wz_mini firmware as booted directly from the SD card. The firmware on the camera would be irrelevant in this particular instance.

As for the "rarely works" comment, I was attempting to be polite to the author of the script and not simply come off like a disgruntled whiner -- which was not my intent. I have great respect and appreciation for you guys and your collective efforts here.

Not much more detail available beyond my initial description of the episode. There were definitely multiple attempts to execute the reboot command hoping to achieve a response.. I was watching the camera waiting for activity after each try and pretty sure I waited the full 90 seconds (or close) each time (no frantic multi-clicking involved) . With no success, I decided to abandon the edited version and 'revert' to the previous version of the file -- which presumably _had_ been edited and presumably saved after executing the "update" command..., or maybe not? I then rebooted from the app which resulted in the loss of wz_mini functionality and a return to the firmware flashed to the cam -- which just happened to be the Wyze RTSP firmware. 

Just to be clear -- the wz_mini.conf _files_ were still there along with a new non-descript wz_mini.conf file. Just no _contents_..., as in "select all" -> "delete" no contents. Completely empty files. Blank. 

Is it possible that the "edited" file was copied/created in name only and not recognized during the "revert" action? If that were the case, could the system possibly have only known of the actual file and "reverted" to what essentially would have been a blank file? I would think the script itself would have at least stayed intact, but there was nothing there at all.

I'm not sure any of this helps. I imagine there are logs somewhere I could check for details?

**Comment by virmaior** [comment]
1. considering that you're describing something that sounds like file corruption, the firmware version might matter. Also, it's a bit inaccurate to suggest that wz_mini runs _instead of_ whatever firmware is on the cam. It runs _on top of_ the firmware from wyze modifying some things using injections.

2. as I stated above, the line that would lead to the reboot was missing so that's why it wasn't happening. (The earliest versions did reboot but then it was suggested to have a separate update and reboot arrangement and the `reboot` line got left out of the update ). All the frantic clicking in the world wouldn't have changed that. Again, it wasn't buggy; it was outright missing.

3. there's nothing in the code that immediately makes sense to me for what you're describing with the empty wz_mini.conf. The revert function literally deletes the current file and renames the one you revert to using ash (wyze's camera's equivalentof bash). So it would not be capable of producing an empty file at that point. 

If it breaks, it's before the revert command. The update process does write a new wz_mini.conf file, which has two halves to how it constructs it. It runs a loop where (1) it copies over from the current wz_mini.conf but (2) replaces lines where there are values from the HTML form.

By ` I would think the script itself would have at least stayed intact, but there was nothing there at all.` , do you mean the wz_mini.conf being blank or do you mean the entire /wz_mini/ directly is getting deleted from your card? If so, that's got to have a different cause.

**Comment by Bob-AOL** [comment]
wz_mini being blank -- all three iterations.

**Reply by virmaior** [reply]
> wz_mini being blank -- all three iterations.

This means that the error occurred during an update and then got copied over ... I'm not really sure how. The first thing that comes to mind  is double-clicking update on the form or something in that vicinity but since I haven't been able to reproduce what you encountered, I don't know how to fix it.

**Reply by Bob-AOL** [reply]
I certainly appreciate your looking into it nonetheless. It seemed like an issue worth bringing up, so I did. 

The interface itself is obviously a very handy feature for systems not concerned about super tight security and/or folks who are simply operating in a GUI environment -- which frequently go hand in hand. I promote it on the Wyze forums for the Linux challenged who still want to take a stab at wz_mini in as easy a way as possible. Having it function as it appears it should only reinforces the integrity of the project.

Thanks again for your work.

**Reply by virmaior** [reply]
thanks. appreciate it.

there's not much we can actually do with the `httpd` available to us for it. (can't really add any sort of https so password protection would be a joke as it has to pass through GET or POST -- not even sure if we have COOKIES access on the server).

I threw in something in the most recent pull request where it won't reboot if there's no wz_mini.conf file. At least that would give someone a chance to try to debug it before killing their camera.

**Comment by Bob-AOL** [comment]
> inaccurate to suggest that wz_mini runs instead of whatever firmware is on the cam.

I never suggested it did.

**Comment by gtxaspec** [comment]
old versions of the firmware, including rtsp, don't reboot properly or at all, intermittently when the reboot command is called, I think this is documented by various folks on the official wyze forums.  due to this, changes on disk may not be sync'ed correctly leading to corruption, as noted in my testing. just fyi
